// allegro-management-api/lib/Command.js
var RouteStore = require('../RouteStore'),
	EventNotifier = require('../EventNotifier'),
	Subscriptions = require('../Subscriptions'),
	CompanyDetailsStore = require('../CompanyDetailsStore');

var Command = function(){
	var routeStore = new RouteStore(),
		eventNotifier = new EventNotifier(),
		subscriptions = new Subscriptions(),
		companyDetailsStore = new CompanyDetailsStore();

	var removeDataCentre = function(dataCentre, callback){
		eventNotifier.raise('remove_data_centre',dataCentre,callback);
	};

	var updateDataCentre = function(oldDataCentre,newDataCentre,callback){
		var updateDataCentre = {
			oldDataCentre: oldDataCentre,
			newDataCentre: newDataCentre
		};
		eventNotifier.raise('update_data_centre',updateDataCentre,callback);
	};

	var addDataCentre = function(dataCentre, callback){
		eventNotifier.raise('add_data_centre',dataCentre,callback);
	};

	var updateRoute = function(route, callback){
		eventNotifier.raise('update_route',route,callback);
	};

	var updateTransitRoute = function(route, callback){
		eventNotifier.raise('update_transit_route',route,callback);
	};

	var updateDevice = function(device, callback){
		eventNotifier.raise('update_device',device,callback);
	};

	var addSubscription = function(subscription,callback){
		subscriptions.subscribe(subscription,callback);
	};

	var removeSubscription = function(subscription,callback){
		subscriptions.unsubscribe(subscription,callback);
	};

	var createManualSP2PQuote = function(quote,callback){
		eventNotifier.raise('manual_quote_created',quote,callback);
	};

	var createManualP2PVLANQuote = function(quote,callback){
		eventNotifier.raise('manual_p2p_vlan_quote_created',quote,callback);
	};

	function createPort(port, callback){
		eventNotifier.raise('manual_port_created',port,callback);
	}

	function createTransitPort(port, callback){
		eventNotifier.raise('manual_transit_port_created',port,callback);
	}

	function createManualTransitQuote(quote, callback){
		eventNotifier.raise('manual_transit_quote_created', quote, callback);
	}

	function addCompany(company, callback){
		eventNotifier.raise('add_company', company, callback);
	}

	function updateTransitPricing(transitPricing, callback){
		eventNotifier.raise('update_transit_pricing', transitPricing, callback);
	}

	function updateTransitDataRate(transitDataRate, callback){
            eventNotifier.raise('update_transit_data_rate', transitDataRate, callback);
    }

    function createTransitVlanTag(transitVlanTag, callback){
    		eventNotifier.raise('add_transit_vlan_tag', transitVlanTag, callback);
    }

	return{
		updateRoute:updateRoute,
		updateTransitRoute:updateTransitRoute,
		removeDataCentre:removeDataCentre,
		addDataCentre:addDataCentre,
		updateDataCentre : updateDataCentre,
		addSubscription: addSubscription,
		removeSubscription: removeSubscription,
		updateDevice: updateDevice,
		createManualSP2PQuote: createManualSP2PQuote,
		createManualP2PVLANQuote: createManualP2PVLANQuote,
		createManualTransitQuote: createManualTransitQuote,
		createPort : createPort,
		createTransitPort : createTransitPort,
		addCompany : addCompany,
		updateTransitPricing : updateTransitPricing,
		updateTransitDataRate : updateTransitDataRate,
		createTransitVlanTag: createTransitVlanTag
	};
	
};

module.exports = new Command();
