// allegro-api/lib/Command.js
var EventHandler = require('../EventHandler');

var Command = function(){
	var eventHandler = new EventHandler();

	
	function createQuote(quote, done){
		eventHandler.quoteRequest(quote, function(){
			done();
		});	
	}

	function createTransitQuote(quote, done){
		eventHandler.transitQuoteRequest(quote, function(){
			done();
		});	
	}

	function purchaseService(quoteReference, done){
		eventHandler.servicePurchase(quoteReference, function(){
			done();
		});		
	}

	function purchaseP2PVlanService(quoteReference, done){
		eventHandler.p2pVlanServicePurchase(quoteReference, function(){
			done();
		});		
	}

	function purchaseTransitService(quoteReference, done){
		eventHandler.transitServicePurchase(quoteReference, function(){
			done();
		});		
	}

	function purchaseNonManualTransitService(quoteReference, vlantag, iprange, done){
		eventHandler.transitNonManualServicePurchase(quoteReference, vlantag, iprange, function(){
			done();
		});		
	}

	function updateService(serviceReference, done){
		eventHandler.serviceUpdate(serviceReference, function(){
			done();
		});		
	}

	function updatep2pVlanService(p2pVlanServiceReference, done){
		eventHandler.p2pVlanServiceUpdate(p2pVlanServiceReference, function(){
			done();
		});		
	}

	function updateTransitService(transitServiceReference, done){
		eventHandler.transitServiceUpdate(transitServiceReference, function(){
			done();
		});		
	}

	function updateCompanyDetails(companyDetails,done){
		eventHandler.updateCompanyDetails(companyDetails, function(){
			done();
		});	
	}

	function addSignedMsa(signedMSA, callback)
	{
		eventHandler.addSignedMsa(signedMSA, function(){
			callback();
		});	
	}

	function addPendingMsa(pendingMSA, callback)
	{
		eventHandler.addPendingMsa(pendingMSA, function(){
			callback();
		});	
	}

	function updateTransitVlanTag(transitVlanTag, callback){
    	eventHandler.updateTransitVlanTag(transitVlanTag, function(){
			callback();
		});	
    }

	return{
		createQuote:createQuote,
		createTransitQuote:createTransitQuote,
		purchaseService : purchaseService,
		purchaseP2PVlanService : purchaseP2PVlanService,
		updateCompanyDetails: updateCompanyDetails,
		updateService : updateService,
		updatep2pVlanService: updatep2pVlanService,
		updateTransitService: updateTransitService,
		purchaseTransitService : purchaseTransitService,
		purchaseNonManualTransitService: purchaseNonManualTransitService,
		addSignedMsa : addSignedMsa,
		addPendingMsa : addPendingMsa,
		updateTransitVlanTag: updateTransitVlanTag
	};
};

module.exports = new Command();
