// allegro-api/lib/Query.js
var DataCentreStore = require('../DataCentreStore'),
	QuoteViewStore = require('../QuoteViewStore'),
	TransitQuoteViewStore = require('../TransitQuoteViewStore'),
	P2PVlanQuoteViewStore = require('../P2PVlanQuoteViewStore'),
	ServicesViewStore = require('../ServicesViewStore'),
	P2PVlanServicesViewStore = require('../P2PVlanServicesViewStore'),
	TransitServicesViewStore = require('../TransitServicesViewStore'),
	TransitProvidersStore = require('../TransitProvidersStoreViewStore'),
	SignedMsaDataStore = require('../SignedMsaDataStore'),
	CompanyDetailsViewStore = require('../CompanyDetailsViewStore'),
	TransitDataRatesStore = require('../TransitDataRatesStore'),
	PendingMsaDataStore = require('../PendingMsaDataStore'),
	TransitVlanTagsViewStore = require('../TransitVlanTagsViewStore');

var Query = function() {
		var dataCentreStore = new DataCentreStore(),
			quoteViewStore = new QuoteViewStore(),
			transitQuoteViewStore = new TransitQuoteViewStore(),
			p2pvlanQuoteViewStore = new P2PVlanQuoteViewStore(),
			singleP2PServicesViewStore = new ServicesViewStore(),
			p2pVlanServicesViewStore = new P2PVlanServicesViewStore(),
			transitServicesViewStore = new TransitServicesViewStore(),
			signedMsaDataStore = new SignedMsaDataStore(),
			transitProvidersStore = new TransitProvidersStore(),
			companyDetailsViewStore = new CompanyDetailsViewStore(),
			transitDataRatesStore = new TransitDataRatesStore(),
			pendingMsaDataStore = new PendingMsaDataStore(),
			transitVlanTagsViewStore = new TransitVlanTagsViewStore();

		var dataCentres = function(callback) {
			dataCentreStore.allDataCentresViewModels(callback);
		};

		var quote = function(quoteId, callback) {
				quoteViewStore.get(quoteId, callback);
			};

		var p2pVlanQuote = function(quoteId, callback) {
				p2pvlanQuoteViewStore.get(quoteId, callback);
			};

		var transitQuote = function(quoteId, callback) {
				transitQuoteViewStore.get(quoteId, callback);
			};

		var services = function(companyAsn, callback) {
				singleP2PServices(companyAsn, callback);
			};

		var service = function(reference, callback) {
				singleP2PService(reference, callback);
			};

		var singleP2PServices = function(companyAsn, callback) {
				singleP2PServicesViewStore.getAll(companyAsn, callback);
			};

		var singleP2PService = function(reference, callback) {
				singleP2PServicesViewStore.get(reference, callback);
			};

		var p2pVlanService = function(reference, callback) {
				p2pVlanServicesViewStore.get(reference, callback);
			};

		var transitService = function(reference, callback) {
			transitServicesViewStore.get(reference, callback);
		};

		function transitServices(ASN, callback) {
			transitServicesViewStore.getAll(ASN, callback);
		};

		function transitProviders(callback){
			transitProvidersStore.all(callback);
		}

		function companyDetails(ASN, callback) {
			companyDetailsViewStore.find(ASN,callback);
		}

		function signedMsa(ASN, callback) {
			signedMsaDataStore.get(ASN, callback);
		}

		function pendingMsa(ASN, callback) {
			pendingMsaDataStore.get(ASN, callback);
		}

		function transitDataRates(callback){
			transitDataRatesStore.getAll(callback);
		}
		function availableVlanTags(providerName, callback) {
			transitVlanTagsViewStore.findAvailable(providerName, callback);
		}

		return {
			dataCentres: dataCentres,
			quote: quote,
			services: services,
			service: service,
			companyDetails: companyDetails,
			transitService: transitService,
			transitServices: transitServices,
			p2pVlanService: p2pVlanService,
			transitQuote: transitQuote,
			p2pVlanQuote: p2pVlanQuote,
			signedMsa: signedMsa,
			pendingMsa: pendingMsa,
			transitProviders : transitProviders,
			transitDataRates : transitDataRates,
			availableVlanTags: availableVlanTags

		};
	};

module.exports = new Query();
